### PR TITLE
Refactoring in prepation of docstring clean-up and extension

### PR DESCRIPTION
**Pull request by mseeger**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py [hunk 1]
```diff
@@ -95,13 +93,13 @@ def __init__(
         config_space: dict,
         metric: str,
         points_to_evaluate: Optional[List[dict]] = None,
-        mode: str = "min",
-        num_min_data_points: int = None,
-        top_n_percent: int = 15,
-        min_bandwidth: float = 1e-3,
-        num_candidates: int = 64,
-        bandwidth_factor: int = 3,
-        random_fraction: float = 0.33,
```
Comment (jgolebiowski) [inline syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py:104]: Same as syne_tune/optimizer/schedulers/searchers/bore/bore.py - I think this is more confusing now - users wont know the default values just by looking at the API but will need to dive into the implementation.

File: syne_tune/optimizer/schedulers/searchers/kde/multi_fidelity_kde_searcher.py [hunk 2]
```diff
@@ -81,16 +79,18 @@ def __init__(
         config_space: Dict,
         metric: str,
         points_to_evaluate: Optional[List[Dict]] = None,
-        mode: str = "min",
-        num_min_data_points: int = None,
-        top_n_percent: int = 15,
-        min_bandwidth: float = 0.1,
-        num_candidates: int = 64,
-        bandwidth_factor: int = 3,
-        random_fraction: float = 0.33,
```
Comment (jgolebiowski) [inline syne_tune/optimizer/schedulers/searchers/kde/multi_fidelity_kde_searcher.py:90]: Same as syne_tune/optimizer/schedulers/searchers/bore/bore.py L40

**Review by jgolebiowski — COMMENTED** [review]
The docstring chnages look good, I have added some comments to the code changes but they are mostly minor.

File: syne_tune/backend/trial_backend.py [hunk 3]
```diff
@@ -126,15 +139,21 @@ def pause_trial(self, trial_id: int, result: Optional[dict] = None):
...
+        """Implements `pause_trial`.
+
+        :param trial_id: ID of trial to pause
+        :param result: Result dict based on which scheduler decided to pause the
+            trial
         """
-        Backend specific operation that pauses the trial.
-        """
-        raise NotImplementedError()
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/backend/trial_backend.py:148]: we've been using abc in the past, but it has quite some issues, for example with serialization, so we removed it. It is one of these things that do not add value.

File: syne_tune/callbacks/tensorboard_callback.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        :param ignore_metrics: Defines which metrics should be ignored. If None,
+            all metrics are reported to Tensorboard.
+        :param target_metric: Defines the metric we aim to optimize. If this
+            argument is set, we report the cumulative optimum of this metric as
+            well as the optimal hyperparameters we have found so far.
+        :param mode: Determined whether we maximize ("max") or minimize ("min")
+            the target metric.
+        """
+        if mode is None:
+            mode = "min"
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:45]: I think it is best practice to define default values of base class methods in the code (so always use = None), otherwise they need to be copy and pasted to subclasses, and this introduces errors. Also, strings are mutable.

File: syne_tune/callbacks/tensorboard_callback.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        target_metric: Optional[str] = None,
+        mode: Optional[str] = None,
+    ):
+        """
+        :param ignore_metrics: Defines which metrics should be ignored. If None,
+            all metrics are reported to Tensorboard.
+        :param target_metric: Defines the metric we aim to optimize. If this
+            argument is set, we report the cumulative optimum of this metric as
+            well as the optimal hyperparameters we have found so far.
+        :param mode: Determined whether we maximize ("max") or minimize ("min")
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:41]: OK.

File: syne_tune/callbacks/tensorboard_callback.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            "curr_best_config": self.curr_best_config,
+            "start_time_stamp": self.start_time_stamp,
+            "iter": self.iter,
+            "mode": self.mode,
+            "target_metric": self.target_metric,
+            "trial_ids": self.trial_ids,
+            "metric_sign": self.metric_sign,
+            "output_path": self.output_path,
+        }
+        return state
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:163]: I don't think this changes anything, as `state` is just been created here. I did not write this code.

File: syne_tune/callbacks/tensorboard_callback.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        )
+
+    def on_tuning_end(self):
+        self.writer.close()
+        logger.info(
+            f"Tensorboard information has been logged at {self.output_path}, to visualize results, run\n"
+            f"tensorboard --logdir {self.output_path}"
+        )
+
+    def __getstate__(self):
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:149]: OK

File: syne_tune/callbacks/tensorboard_callback.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            "iter": self.iter,
+            "mode": self.mode,
+            "target_metric": self.target_metric,
+            "trial_ids": self.trial_ids,
+            "metric_sign": self.metric_sign,
+            "output_path": self.output_path,
+        }
+        return state
+
+    def __setstate__(self, state):
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:165]: That sounds a bit unsafe to me.

File: syne_tune/experiments.py [hunk 9]
```diff
@@ -184,79 +194,139 @@ def load_experiment(
...
+
+    :param path_filter: If passed then only experiments whose path matching
+        the filter are kept. This allows rapid filtering in the presence of many
+        experiments.
+    :param root: Root path for experiment results. Default is
+        `experiment_path()`
+    :return: Dictionary from tuner name to metadata dict
     """
-    res = {}
+    path_filter = _impute_path_filter(path_filter)
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/experiments.py:233]: Use of these helper functions seems cleaner to me, as this is needed several times. Also, PyCharm tells me not to use lambda, but define the function directly.

[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/experiments.py:233]: I actually do not like lambda functions a lot, try not to use them.

File: syne_tune/experiments.py [hunk 10]
```diff
@@ -184,79 +194,139 @@ def load_experiment(
...
+    return pfilter
+
+
+def _impute_experiment_filter(efilter: ExperimentFilter) -> ExperimentFilter:
+    if efilter is None:
+
+        def efilter(result) -> bool:
+            return True
+
+    return efilter
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/experiments.py:218]: OK

File: syne_tune/experiments.py [hunk 11]
```diff
@@ -184,79 +194,139 @@ def load_experiment(
...
+    :param path_filter: If passed then only experiments whose path matching
+        the filter are kept. This allows rapid filtering in the presence of many
+        experiments.
+    :param experiment_filter: Filter on `ExperimentResult`
+    :param root: Root path for experiment results. Default is
+        `experiment_path()`
+    :param load_tuner: Whether to load the tuner in addition to metadata and results
+    :return: List of `ExperimentResult` objects
+    """
+    path_filter = _impute_path_filter(path_filter)
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/experiments.py:272]: OK

File: syne_tune/experiments.py [hunk 12]
```diff
@@ -184,79 +194,139 @@ def load_experiment(
...
+    if experiment_filter is None:
+
+        def experiment_filter(result) -> bool:
+            return True
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/experiments.py:276]: OK

File: syne_tune/optimizer/schedulers/fifo.py [hunk 13]
```diff
@@ -376,32 +338,26 @@ def on_trial_result(self, trial: Trial, result: dict) -> str:
         self._check_result(result)
         trial_id = str(trial.trial_id)
         trial_decision = SchedulerDecision.CONTINUE
-        if len(result) == 0:
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/optimizer/schedulers/fifo.py:379]: This cannot happen. It is some old code, when the scheduler was used with another backend (AutoGluon), where this could happen.

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py [hunk 14]
```diff
@@ -37,19 +36,18 @@ def __init__(
...
-        points_to_evaluate=None,
-        mode: str = "max",
-        gamma: float = 0.25,
-        calibrate: bool = False,
-        classifier: str = "xgboost",
-        acq_optimizer: str = "rs",
-        feval_acq: int = 500,
-        random_prob: float = 0.0,
-        init_random: int = 6,
-        classifier_kwargs: dict = None,
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bore/bore.py:49]: That is best practice for methods of base classes. We can state the default values in the docstring.

Also, if the assignments of default values happen just at the start of a method, a user looking at the code will have no more trouble seeing them.

File: syne_tune/optimizer/schedulers/searchers/bore/multi_fidelity_bore.py [hunk 15]
```diff
@@ -10,72 +10,63 @@
...
+        gamma: Optional[float] = None,
+        calibrate: Optional[bool] = None,
+        classifier: Optional[str] = None,
+        acq_optimizer: Optional[str] = None,
+        feval_acq: Optional[int] = None,
+        random_prob: Optional[float] = None,
+        init_random: Optional[int] = None,
+        classifier_kwargs: Optional[dict] = None,
         resource_attr: str = "epoch",
-        **kwargs
```
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bore/multi_fidelity_bore.py:71]: That is the whole point in the subclass. If something depends to the base class, you should *have* to look there.
Copy and paste is not the solution.

I am all for mentioning default values in the docstring. This is apparently done in Google style, but not in reST we are using. I am not an expert on this.

It is even worse if we end up relying on copy and paste, and then suddenly somebody changes a def value in a derived class. Is this a copy mistake, or is it intended?

**Comment by mseeger** [comment]
OK, I think I have addressed concerns. About default values: We can include them in docstrings, but I am against copy
and paste them to subclasses. It is best practice to use `= None` and deal with them in the code of the base class.

File: syne_tune/backend/trial_backend.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/backend/trial_backend.py:148]: I see, if you have seen problems with serialization of ABC's classes then lets leave it like this.

File: syne_tune/callbacks/tensorboard_callback.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/callbacks/tensorboard_callback.py:45]: I see your point about inheritance, as long as we keep the default in the docstrings then I am ok with that. 

I think the best practice are arguable, see for example an implementation of Adam in tensorflow https://github.com/keras-team/keras/blob/v2.10.0/keras/optimizers/optimizer_v2/adam.py#L26-L270 or sgd in Pytorch https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD which are(IMHO) a relatively well written library 

As a note, strings in python are immutable, see second sentence of https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str

File: syne_tune/callbacks/tensorboard_callback.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/callbacks/tensorboard_callback.py:157]: Good point, I was thinking about the version with 
```
del state['writer']
return state
```
which is confusing. The current version is good.

File: syne_tune/callbacks/tensorboard_callback.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/callbacks/tensorboard_callback.py:159]: I see the concern, in this case I am happy with the current state.

File: syne_tune/optimizer/schedulers/fifo.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/optimizer/schedulers/fifo.py:379]: Ok makes sense, thanks for clearing it up.

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/optimizer/schedulers/searchers/bore/bore.py:49]: I see, as long as the default values are in the docstring it makes sense to me.

File: syne_tune/optimizer/schedulers/searchers/bore/multi_fidelity_bore.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jgolebiowski) [reply syne_tune/optimizer/schedulers/searchers/bore/multi_fidelity_bore.py:71]: I see, as long as the default values are in the docstring it makes sense to me.

File: syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py (hunk 1 above)
Comment (jgolebiowski) [reply syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py:104]: NOTE: as long as the default values are in the docstring it makes sense to me.

**Comment by jgolebiowski** [comment]
Good changes - I agree with the current implementation if we add defaults into docstrings for classes that are inherited from.

File: syne_tune/callbacks/tensorboard_callback.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/callbacks/tensorboard_callback.py:45]: I just checked the sphinx docs, and they recommend to mention type and default in the docstring, something we are not right now doing. In the second PR, I will do that for all classes which are user-facing. This way, we get all of that info also in the API docs.

File: syne_tune/optimizer/schedulers/searchers/bore/bore.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mseeger) [reply syne_tune/optimizer/schedulers/searchers/bore/bore.py:49]: OK, will do that in second PR, where docstrings are updated.

**Review by jgolebiowski — APPROVED** [review]
Docsting changes will come in the 2nd PR as agreed offline.
